proxy/src/bind.rs

```diff
@@ -306,6 +320,50 @@ where
     }
 }
 
+// ===== impl Binding =====
+
+impl<B: tower_h2::Body + 'static> tower::Service for Binding<B> {
+    type Request = <Stack<B> as tower::Service>::Request;
+    type Response = <Stack<B> as tower::Service>::Response;
+    type Error = <Stack<B> as tower::Service>::Error;
+    type Future = <Stack<B> as tower::Service>::Future;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        match *self {
+            // A service is already bound, so poll its readiness.
+            Binding::EagerReusable(ref mut svc) |
+            Binding::LazyOneoff { next: Some(ref mut svc), .. } => svc.poll_ready(),
+
+            // If no stack has been bound, bind it now so that its readiness can be
+            // checked. Store it so it can be consumed to dispatch the next request.
+            Binding::LazyOneoff { ref endpoint, ref protocol, ref bind, ref mut next } => {
+                let mut svc = bind.bind_stack(endpoint, protocol);
+                let ready = svc.poll_ready()?;
+                *next = Some(svc);
+                Ok(ready)
+            }
+        }
+    }
+
+    fn call(&mut self, request: Self::Request) -> Self::Future {
+        match *self {
+            // Dispatch the request immediately.
+            Binding::EagerReusable(ref mut svc) => svc.call(request),
+
+            // `poll_ready` was not called, so attempt to bind a new stack and dispatch
+            // the request immediately.
+            Binding::LazyOneoff { ref endpoint, ref protocol, ref bind, next: None } => {
+                bind.bind_stack(endpoint, protocol).call(request)
+            }
+
+            // A service has already been bound, so consume it and send the request to it.
+            Binding::LazyOneoff { ref mut next, .. } => {
+                next.take().unwrap().call(request)
```

Comment:
Does this mean that a `LazyOneoff` would actually be reusable, just as long as it binds a new stack for each request?

Reply:
Yeah, perhaps the naming isn't as clear as it could be...

the two variants are:
1. we have bound a service early and can reuse it indefinitely.
2. we have to bind a new service for each request.

`LazyOneoff` refers to the second case.  Because we want to honor the `poll_ready` API, we may still bind a service before `call` is invoked, and we need to take care to consume it here.